### PR TITLE
Support defining image paths for HelmReleases

### PR DESCRIPTION
**Pull request by hiddeco**
This commit adds support for supplying image paths for `HelmRelease`
resources through annotations, and introduces a new alias for image
tag filters.

For an image mapping to be valid, the image path _must_ be defined
and _may_ be supplemented by a tag and/or registry path. In case
the combination of resolved values does not result in a valid image
ref, the mapping will be ignored.

**Supported annotations:**

`registry.fluxcd.io/<alias>: <registry path>`
`repository.fluxcd.io/<alias>: <image path>`
`tag.fluxcd.io/<alias>: <tag path>`
`filter.fluxcd.io/<alias>: <tag filter>`

Image mappings are composed by `<alias>`.

**A valid annotation mapping is for example:**

```yaml
annotations:
  registry.fluxcd.io/container: some.custom.registry
  repository.fluxcd.io/container: some.custom.image
  tag.fluxcd.io/container: some.custom.tag
  filter.fluxcd.io/container: glob: master-*

  repository.fluxcd.io/anotherone: depth.knows.no.limits.image
```

which would map the image for the following `values` structure:

```yaml
values:
  some:
    custom:
      registry: docker.io
      image: fluxcd/flux-prerelease
      tag: master-abcdefg
  depth:
    knows:
      no:
        limits:
          image: fluxcd/helm-operator:0.10.0
```

This PR is in need of ~https://github.com/squaremo/kubeyaml/pull/11~ https://github.com/squaremo/kubeyaml/pull/14 to persist the actual changes to git.

Fixes: #1262

### CODE REVIEW

**Comment by 2opremio** [comment]
Where, `container` a parameter, right? (i.e. does the scheme support multiple containers?) 

The use of subdomains (as opposed to e.g. a path) to denote what you are specifying seems a bit funky, but we probably can't do better due to the restrictions of the annotation syntax imposed by k8s.

Maybe we could add another subdomain to make it more clear? e.g.
`tag.pathspec.flux.weave.works`?

**Comment by hiddeco** [comment]
> Where, `container` a parameter, right? (i.e. does the scheme support multiple containers?)

`container` in this case is just an alias for your collection of mappings (it represents one image definition), it can be anything and multiple mappings can be defined.

> The use of subdomains (as opposed to e.g. a path) to denote what you are specifying seems a bit funky, but we probably can't do better due to the restrictions of the annotation syntax imposed by k8s.

Correct, this is the most Kubernetes native approach to defining the mapping and is conform their specifications for annotations.

> Maybe we could add another subdomain to make it more clear? e.g.
`tag.pathspec.flux.weave.works`?

Can not decide if the clarity improvement would weight up to the annoyance of a really long annotation key.

File: cluster/kubernetes/resource/fluxhelmrelease.go [hunk 1]
```diff
@@ -220,18 +327,109 @@ func interpretAsImage(m mapper) (image.Ref, ImageSetter, bool) {
...
+	}
+
+	imageValue := v.Path(cim.Repository).Data()
+	if img, ok := imageValue.(string); ok {
+		if cim.Registry == "" && cim.Tag == "" {
+			if imgRef, err := image.ParseRef(img); err == nil {
+				return imgRef, func(ref image.Ref) {
+					v.SetP(ref.String(), cim.Repository)
+				}, true
+			}
```
Comment (squaremo) [inline cluster/kubernetes/resource/fluxhelmrelease.go]: If this fails, I'm not sure it's worth continuing -- adding a registry and/or tag to an invalid image can't make it valid.

Comment (squaremo) [reply cluster/kubernetes/resource/fluxhelmrelease.go]: (it does get checked again later, so it's safe as it is I believe)

File: cluster/kubernetes/resource/fluxhelmrelease.go [hunk 2]
```diff
@@ -220,18 +327,109 @@ func interpretAsImage(m mapper) (image.Ref, ImageSetter, bool) {
...
+						v.SetP(ref.Tag, cim.Tag)
+					}, true
+				}
+			}
+		}
+	}
+
+	return image.Ref{}, nil, false
+}
+
```
Comment (squaremo) [inline cluster/kubernetes/resource/fluxhelmrelease.go]: :heavy_check_mark: although I should mention it took me quite a lot of staring to get to this point :-)

File: cluster/kubernetes/resource/fluxhelmrelease.go [hunk 3]
```diff
@@ -220,18 +327,109 @@ func interpretAsImage(m mapper) (image.Ref, ImageSetter, bool) {
...
 // Containers returns the containers that are defined in the
 // FluxHelmRelease.
 func (fhr FluxHelmRelease) Containers() []resource.Container {
 	var containers []resource.Container
-	// If there's an error in interpreting, return what we have.
-	_ = FindFluxHelmReleaseContainers(fhr.Spec.Values, func(container string, image image.Ref, _ ImageSetter) error {
+
+	containerSetter := func(container string, image image.Ref, _ ImageSetter) error {
```
Comment (squaremo) [inline cluster/kubernetes/resource/fluxhelmrelease.go]: If you're giving this a name, `addContainer` might be better -- the implied analogy to `imageSetter` doesn't really work.

File: cluster/kubernetes/resource/resource.go [hunk 4]
```diff
@@ -198,7 +205,11 @@ func unmarshalKind(base baseObject, bytes []byte) (KubeManifest, error) {
 		return &list, nil
 	case base.Kind == "FluxHelmRelease" || base.Kind == "HelmRelease":
 		var fhr = FluxHelmRelease{baseObject: base}
-		if err := yaml.Unmarshal(bytes, &fhr); err != nil {
+		// NB: workaround for go-yaml/yaml/issues/139
+		// By using github.com/ghodss/yaml to unmarshal HelmReleases.
+		// This effectively results in all keys of `Value`s being strings
+		// and not interface{}.
+		if err := jsonyaml.Unmarshal(bytes, &fhr); err != nil {
```
Comment (squaremo) [inline cluster/kubernetes/resource/resource.go]: Does this mean the `mapper` stuff above can be removed?

File: cluster/kubernetes/resource/fluxhelmrelease.go [hunk 5]
```diff
@@ -220,18 +327,109 @@ func interpretAsImage(m mapper) (image.Ref, ImageSetter, bool) {
...
+
+func interpretMappedContainerImage(values map[string]interface{}, cim ContainerImageMap) (image.Ref, ImageSetter, bool) {
+	v, err := gabs.Consume(values)
+	if err != nil {
+		return image.Ref{}, nil, false
+	}
+
+	imageValue := v.Path(cim.Repository).Data()
+	if img, ok := imageValue.(string); ok {
+		if cim.Registry == "" && cim.Tag == "" {
```
[the remark this answers is not in the document]
Comment (hiddeco) [reply cluster/kubernetes/resource/fluxhelmrelease.go]: That will definitely make it easier to understand. I added the utilities around `ContainerImageMap` much later and did not rethink if I could also improve this. Thanks for the suggestion.

File: cluster/kubernetes/manifests.go [hunk 6]
```diff
@@ -133,7 +133,34 @@ func (m *manifests) ParseManifest(def []byte, source string) (map[string]resourc
...
+	// in to a valid image ref), fallback to the default behaviour.
+	//
+	// NB: we do this here and not in e.g. the `resource` package, to ensure
+	// everything _outside_ this package only knows about Kubernetes native
+	// containers, and not the dot notation YAML paths we invented for custom
+	// Helm value structures.
+	if hr, ok := res.(*kresource.FluxHelmRelease); ok {
+		if paths, err := hr.GetContainerImageMap(container); err == nil {
+			def, err := updateWorkloadImagePaths(def, id, paths, image)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (hiddeco) [reply cluster/kubernetes/manifests.go]: Agreed.

File: cluster/kubernetes/resource/resource.go (hunk 4 above)
Comment (hiddeco) [reply cluster/kubernetes/resource/resource.go]: Above where? Quite a lot of 'mapper' things on this page...

**Pull request merged**